superduperdb/ext/llm/train.py

```diff
@@ -0,0 +1,299 @@
+        default=512,
+        metadata={
+            "help": "Maximum sequence length. Sequences will be right padded (and possibly truncated)."
+        },
+    )
+
+
+@dataclass
+class DataArguments:
+    data_name: str = field(
+        default="c-s-ale/alpaca-gpt4-data-zh",
+        metadata={"help": "dataset name"},
+    )
+
+
+@dataclass
+class LoraArguments:
+    lora_r: int = 8
+    lora_alpha: int = 16
+    lora_dropout: float = 0.05
+    lora_target_modules: Optional[List[str]] = None
+    lora_weight_path: str = ""
+    lora_bias: str = "none"
+    q_lora: bool = False
+    bits: Optional[int] = None
+
+
+instruction_template = {
+    "description": "Template used by Alpaca-LoRA.",
+    "prompt_input": "Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:\n",
+    "prompt_no_input": "Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\n{instruction}\n\n### Response:\n",
+    "response_split": "### Response:",
+}
+
+
+def create_dataset(tokenizer, data_args):
+    def process_dataset(example):
+        instruction = example["instruction"]
+        input = example.get("input", None)
+        output = example.get("output", None)
+
+        if input is not None:
+            prompt = instruction_template["prompt_input"].format(
+                instruction=instruction, input=input
+            )
+
+        else:
+            prompt = instruction_template["prompt_no_input"].format(
+                instruction=instruction
+            )
+
+        if output is not None:
+            prompt = prompt + output + tokenizer.eos_token
+        result = tokenizer(
+            prompt,
+            truncation=True,
+            max_length=tokenizer.model_max_length,
+            padding="max_length",
+        )
+        result["labels"] = result["input_ids"].copy()
+        return result
+
+    dataset = load_dataset(data_args.data_name, split="train[:3000]")
+
+    dataset = dataset.train_test_split(
+        test_size=0.1,
+        seed=42,
+    )
+    train_dataset = dataset["train"].map(process_dataset)
+    eval_dataset = dataset["test"].map(process_dataset)
+    return train_dataset, eval_dataset
+
+
+def find_all_linear_names(args, model):
+    cls = (
+        bnb.nn.Linear4bit
+        if args.bits == 4
+        else (bnb.nn.Linear8bitLt if args.bits == 8 else torch.nn.Linear)
+    )
+    lora_module_names = set()
+    for name, module in model.named_modules():
+        if isinstance(module, cls):
+            names = name.split(".")
+            lora_module_names.add(names[0] if len(names) == 1 else names[-1])
+
+    if "lm_head" in lora_module_names:  # needed for 16-bit
+        lora_module_names.remove("lm_head")
+    return list(lora_module_names)
+
+
+@dataclass
+class LLMTrainer:
```

Comment:
Is it possible for us to stay with the `model.fit` formalism? This might require some refactoring of your code.

The aim is to stay as much as possible in the old `sklearn` world of:

- `model.predict`
- `model.fit`

Reply:
Sorry, This is an unfinished draft, there will be significant changes. I will close it for now and reopen it later.

Reply:
Ok, no worries.